### PR TITLE
fix(expectations): allow custom matchers to return a promise when act…

### PR DESCRIPTION
**Pull request by sjurba**
…ual is not a promise

See angular/protractor#2964

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -169,7 +169,7 @@ jasmine.Expectation.prototype.wrapCompare = function(name, matcherFactory) {
 
     if (!webdriver.promise.isPromise(expectation.actual) &&
         !webdriver.promise.isPromise(expected)) {
```
Comment (sjelin) [inline index.js]: Why even have the no promises case if we're going to pass everything to `compare` anyway?

File: index.js [hunk 2]
```diff
@@ -191,36 +191,42 @@ jasmine.Expectation.prototype.wrapCompare = function(name, matcherFactory) {
 
       var result = matcherCompare.apply(null, args);
 
-      return webdriver.promise.when(result.pass).then(function(pass) {
-        var message = '';
+      if (webdriver.promise.isPromise(result.pass)) {
```
Comment (sjelin) [inline index.js]: Why this change?  Even if `result.pass` weren't a promise `webdriver.promise.when` should still work shouldn't it?

File: index.js (hunk 1 above)
Comment (sjurba) [inline index.js]: Because once you call when(), you force the call to become async. With this check the it doesn't become async unless it needs to. This was the original reason this code changed, and broke, in the first place. See https://github.com/angular/protractor/issues/2894

File: index.js (hunk 2 above)
Comment (sjurba) [inline index.js]: Same as above...

**Comment by sjelin** [comment]
LGTM, @juliemr should probably take a look though

File: index.js [hunk 3]
```diff
@@ -169,7 +169,7 @@ jasmine.Expectation.prototype.wrapCompare = function(name, matcherFactory) {
 
     if (!webdriver.promise.isPromise(expectation.actual) &&
         !webdriver.promise.isPromise(expected)) {
-      originalWrapCompare(name, matcherFactory).apply(this, arguments);
+      compare(expectation.actual, expected);
```
Comment (juliemr) [inline index.js]: I believe we can remove `originalWrapCompare` above now, since it's now  unused.

**Comment by juliemr** [comment]
LGTM after cleanup.

**Comment by sjurba** [comment]
@juliemr done.

**Comment by juliemr** [comment]
Thanks! We'll push this into a release when we do the next protractor version bump.

**Comment by laseelan** [comment]
you mean the next jasminewd2 version bump ?

**Comment by sjurba** [comment]
@juliemr Any idea when we can expect a new protractor version bump? Was this forgotten for the 3.2 releases?

**Comment by juliemr** [comment]
This was, sorry about that. Version bump is now in master (https://github.com/angular/protractor/commit/b4d1664141b609c5e5790108e3003fe777c248ca) and will be out in the next release.
